// ng-whiteboard/esm2015/lib/ng-whiteboard.types.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ng-whiteboard.types.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class WhiteboardOptions {
    constructor() {
        this.color = '#000000';
        this.backgroundColor = '#ffffff';
        this.size = '5px';
        this.linejoin = 'round';
        this.linecap = 'round';
    }
}
if (false) {
    /** @type {?} */
    WhiteboardOptions.prototype.color;
    /** @type {?} */
    WhiteboardOptions.prototype.backgroundColor;
    /** @type {?} */
    WhiteboardOptions.prototype.size;
    /** @type {?} */
    WhiteboardOptions.prototype.linejoin;
    /** @type {?} */
    WhiteboardOptions.prototype.linecap;
}
/**
 * @record
 */
export function ActionStack() { }
if (false) {
    /** @type {?} */
    ActionStack.prototype.type;
    /** @type {?|undefined} */
    ActionStack.prototype.line;
    /** @type {?|undefined} */
    ActionStack.prototype.image;
}
/** @enum {number} */
const ActionType = {
    Line: 0,
    Image: 1,
};
export { ActionType };
ActionType[ActionType.Line] = 'Line';
ActionType[ActionType.Image] = 'Image';
/** @enum {string} */
const FormatType = {
    Png: "png",
    Jpeg: "jpeg",
    Svg: "svg",
    Base64: "base64",
};
export { FormatType };
